src/utils/recorder_events.jl

```diff
@@ -0,0 +1,290 @@
+abstract type AbstractRecorderEvent end
+
+get_common(event::AbstractRecorderEvent) = event.common
+get_name(event::AbstractRecorderEvent) = get_name(get_common(event))
+get_timestamp(event::AbstractRecorderEvent) = get_timestamp(get_common(event))
+
+struct RecorderEventCommon
+    name::String
+    timestamp::Dates.DateTime
+end
+
+function RecorderEventCommon(name::AbstractString)
+    return RecorderEventCommon(name, Dates.now())
+end
+
+get_name(event::RecorderEventCommon) = event.name
+get_timestamp(event::RecorderEventCommon) = event.timestamp
+
+function serialize(event::T) where {T <: AbstractRecorderEvent}
+    data = Dict{Symbol, Any}()
+
+    for (fieldname, fieldtype) in zip(fieldnames(T), fieldtypes(T))
+        if fieldtype <: RecorderEventCommon
+            data[:name] = get_name(event)
+            data[:timestamp] = get_timestamp(event)
+        else
+            data[fieldname] = getfield(event, fieldname)
+        end
+    end
+
+    return JSON.json(data)
+end
+
+"""
+    deserialize(::Type{T}, text::AbstractString) where T <: AbstractRecorderEvent
+
+Deserialize an event from raw text. Will throw an exception if the text does not contain
+a serialized version of T.
+"""
+function deserialize(::Type{T}, text::AbstractString) where {T <: AbstractRecorderEvent}
+    data = JSON.parse(text)
+    name = pop!(data, "name")
+    timestamp = Dates.DateTime(pop!(data, "timestamp"))
+    common = RecorderEventCommon(name, timestamp)
+    vals = []
+
+    for (fieldname, fieldtype) in zip(fieldnames(T), fieldtypes(T))
+        if fieldtype <: RecorderEventCommon
+            push!(vals, common)
+        else
+            push!(vals, data[string(fieldname)])
+        end
+    end
+
+    return T(vals...)
+end
+
+"""
+Records user-defined events in JSON format.
+"""
+mutable struct Recorder
+    name::Symbol
+    io::IO
+end
+
+"""
+    Recorder(
+        name::Symbol;
+        io::Union{Nothing, IO} = nothing,
+        mode = "w",
+        directory = ".",
+    )
+
+Construct a Recorder.
+
+# Arguments
+- `name::Symbol`: name of recorder
+- `io::Union{Nothing, IO}`:  If nothing, record events in a file using name.
+- `mode = "w"`:  Only used when io is nothing.
+- `directory = "."`:  Only used when io is nothing.
+"""
+function Recorder(
+    name::Symbol;
+    io::Union{Nothing, IO} = nothing,
+    mode = "w",
+    directory = ".",
+)
+    if isnothing(io)
+        filename = joinpath(directory, string(name) * "_recorder.log")
```

Comment:
The default behavior is store events for each recorder in its own file.  I considered storing all events for all loggers in one file.  My initial thought was that this provides the user better control. Also, it would prevent files with tons of events from crowding out files with few events.  We could add file rotation based on size in the future, if needed.

Feel free to make a different argument.